regex/src/main/scala/regex/Regex.scala

```diff
@@ -1,35 +1,70 @@
 package regex
 
 /** *****************************************************************************
-  * Regular Languages
-  *
-  * data structure definitions for regular languages
-  */
+  * Regular Languages
+  *
+  * data structure definitions for regular languages
+  */
 
-// Add your definitions here
+trait RegularLanguage
 
-/** *****************************************************************************
-  * Derivatives
-  *
-  * Fill in the function definitions below
-  */
+case object Empty extends RegularLanguage
+
+case object Epsilon extends RegularLanguage
+
+case class Character(c: Char) extends RegularLanguage
+
+case class Union(l1: RegularLanguage, l2: RegularLanguage) extends RegularLanguage
+
+case class Concat(l1: RegularLanguage, l2: RegularLanguage) extends RegularLanguage
+
+case class Star(l: RegularLanguage) extends RegularLanguage
 
 /** Simplifies a regular language */
-def simplify(lang: RegularLanguage): RegularLanguage = ???
+def simplify(lang: RegularLanguage): RegularLanguage = lang match {
+  case Concat(Epsilon, sublang) => simplify(sublang)
+  case Concat(sublang, Epsilon) => simplify(sublang)
+  case Concat(Empty, sublang) => simplify(Empty)
+  case Concat(sublang, Empty) => simplify(Empty)
+  case Concat(sublang1, sublang2) => Concat(simplify(sublang1), simplify(sublang2))
+  case Union(Empty, sublang) => simplify(sublang)
+  case Union(sublang, Empty) => simplify(sublang)
+  case Union(sublang1, sublang2) => Union(simplify(sublang1), simplify(sublang2))
+  case Star(Epsilon) => Epsilon
+  case Star(Empty) => Empty
+  case Star(sublang) => Star(simplify(sublang))
+  case _ => lang
+}
+
```

Comment:
I pattern matched a bit differently - I thought this version was impossible, but that could have been because I implemented most of it before I made Empty and Epsilon case objects. This way is more like the pattern matching I've seen before.